libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,216 @@
+	else
+		return gasConsumption.value;
+}
+/// @returns an estimation of the code size in bytes needed for the AssemblyItems in @a _itemRange.
+template<typename RangeType>
+uint64_t codeSize(RangeType&& _itemRange)
+{
+	// TODO: what is the best choice for the ``_addressLength`` argument here?
+	return ranges::accumulate(std::forward<RangeType>(_itemRange) | ranges::views::transform(
+			[](auto const& _item) { return _item.bytesRequired(2); }
+	), 0u);
+}
+}
+
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
+{
+	AssemblyItem exitJump = _block.items.back();
+
+	if (
+		_jump.getJumpType() == AssemblyItem::JumpType::IntoFunction &&
+		exitJump.getJumpType() == AssemblyItem::JumpType::OutOfFunction
+	)
+	{
+		exitJump.setJumpType(AssemblyItem::JumpType::Ordinary);
+
+		// Accumulate size of the inline candidate block in bytes (without the return jump).
+		uint64_t functionBodySize = codeSize(ranges::views::drop_last(_block.items, 1));
+
+		// Use the number of push tags as approximation of the average number of calls to the function per run.
+		uint64_t numberOfCalls = _block.pushTagCount;
+		// Also use the number of push tags as approximation of the number of call sites to the function.
+		uint64_t numberOfCallSites = _block.pushTagCount;
+
+		static AssemblyItems const uninlinedCallSitePattern = {
+			AssemblyItem{PushTag},
+			AssemblyItem{PushTag},
+			AssemblyItem{Instruction::JUMP},
+			AssemblyItem{Tag}
+		};
+		static AssemblyItems const uninlinedJumpSitePattern = {
+			AssemblyItem{Tag},
+			// Actual function body of size functionBodySize. Handled separately below.
+			AssemblyItem{Instruction::JUMP}
+		};
+
+		// Both the call site and jump site pattern is executed for each call.
+		// Since the function body has to be executed equally often both with and without inlining,
+		// it can be ignored.
+		bigint uninlinedExecutionCost = numberOfCalls * (
+			executionCost(uninlinedCallSitePattern, m_evmVersion) +
+			executionCost(uninlinedJumpSitePattern, m_evmVersion)
+		);
+		// Each call site deposits the call site pattern, whereas the jump site pattern and the function itself are deposited once.
+		bigint uninlinedDepositCost = GasMeter::dataGas(
+			numberOfCallSites * codeSize(uninlinedCallSitePattern) +
+			codeSize(uninlinedJumpSitePattern) +
+			functionBodySize,
+			m_isCreation,
+			m_evmVersion
+		);
+		// When inlining the execution cost beyond the actual function execution is zero,
+		// but for each call site a copy of the function is deposited.
+		bigint inlinedDepositCost = GasMeter::dataGas(numberOfCallSites * functionBodySize, m_isCreation, m_evmVersion);
+
+		// If the estimated runtime cost over the lifetime of the contract plus the deposit cost in the uninlined case
+		// exceed the inlined deposit costs, it is beneficial to inline.
+		if (bigint(m_runs) * uninlinedExecutionCost + uninlinedDepositCost > inlinedDepositCost)
```

Comment:
Does this assume we can always delete the function if it is inlined?

Reply:
Yes... which is of course wrong in general... I could also calculate under the assumption that we *can't* remove it. It would probably also be possible to be sure about it, but only for "right now" and not for "maybe after further optimization cycles"...

Reply:
Can you take references from the surrounding assembly into account? That will probably not be removable since it is a dynamic jump.

Reply:
Yeah... let me see if I can... with "surrounding assembly" you mean the assembly the current assembly is a subassembly of, right :-)?

Reply:
Ah, ok, there is ``_tagsReferencedFromOutside`` available, I'll just have to pass that through, nice! I'll do that then!